codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -272,6 +275,13 @@ private boolean readClientPrefaceString(ByteBuf in) throws Http2Exception {
             if (bytesRead == 0 || !ByteBufUtil.equals(in, in.readerIndex(),
                                                       clientPrefaceString, clientPrefaceString.readerIndex(),
                                                       bytesRead)) {
+                String http1Request =
```

Comment:
also you mention that this is careful not to include any header data ... how is that accomplished? Should we more explicitly look for the CRLF to ensure we only include the request-line? 

```diff
diff --git a/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java b/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java
index dfe677d..3b4dbf8 100644
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java
@@ -16,6 +16,7 @@ package io.netty.handler.codec.http2;
 
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelHandlerContext;
@@ -26,6 +27,7 @@ import io.netty.handler.codec.http.HttpResponseStatus;
 import io.netty.handler.codec.http2.Http2Exception.CompositeStreamException;
 import io.netty.handler.codec.http2.Http2Exception.StreamException;
 import io.netty.util.AsciiString;
+import io.netty.util.ByteProcessor;
 import io.netty.util.CharsetUtil;
 import io.netty.util.concurrent.ScheduledFuture;
 import io.netty.util.internal.UnstableApi;
@@ -198,6 +200,9 @@ public class Http2ConnectionHandler extends ByteToMessageDecoder implements Http
         }
     }
 
+    private static final ByteBuf HTTP_1_X_PREFIX_BUF = Unpooled.unreleasableBuffer(Unpooled.wrappedBuffer(
+            new byte[] {'H', 'T', 'T', 'P', '/', '1', '.'})).asReadOnly();
+
     private final class PrefaceDecoder extends BaseDecoder {
         private ByteBuf clientPrefaceString;
         private boolean prefaceSent;
@@ -275,12 +280,16 @@ public class Http2ConnectionHandler extends ByteToMessageDecoder implements Http
             if (bytesRead == 0 || !ByteBufUtil.equals(in, in.readerIndex(),
                                                       clientPrefaceString, clientPrefaceString.readerIndex(),
                                                       bytesRead)) {
-                String http1Request =
-                    in.toString(in.readerIndex(), Math.min(in.readableBytes(), 1024), CharsetUtil.US_ASCII);
-                int http1Index;
-                if ((http1Index = http1Request.indexOf("HTTP/1.")) != -1) {
+                if (ByteBufUtil.equals(HTTP_1_X_PREFIX_BUF, HTTP_1_X_PREFIX_BUF.readerIndex(),
+                                   in, in.readerIndex(), HTTP_1_X_PREFIX_BUF.readableBytes())) {
+                    final int maxLength = 1024;
+                    int endIndex = in.forEachByte(in.readerIndex(), Math.min(in.readableBytes(), maxLength),
+                                                  ByteProcessor.FIND_CRLF);
+                    if (endIndex < 0) {
+                        endIndex = in.readerIndex() + maxLength;
+                    }
                     throw connectionError(PROTOCOL_ERROR, "Unexpected HTTP/1.x request: %s",
-                                                          http1Request.substring(0, http1Index));
+                            in.toString(in.readerIndex(), endIndex - in.readerIndex(), CharsetUtil.US_ASCII));
                 }
                 String receivedBytes = hexDump(in, in.readerIndex(),
                                                min(in.readableBytes(), clientPrefaceString.readableBytes()));

```